mite_selenium/__init__.py

```diff
@@ -51,32 +56,119 @@ def start(self):
     def stop(self):
         self._context.browser.close()
 
-
-def send_selenium_times(context):
-    """Maybe would probably need more error handling
-    if somebody tried to use with an incompatible browser"""
-    if context.browser.capabilities['browserName'] != 'chrome':
-        logger.warning(
-            'Timing metrics in %s are not currently supported',
-            context.browser.capabilities['browserName'],
-        )
-        return
-
-    timings = context.browser.execute_script(
-        'return performance.getEntriesByType("navigation")'
-    )[0]
-    context.send(
-        "http_selenium_metrics",
-        dom_complete=timings["domComplete"],
-        dom_interactive=timings["domInteractive"],
-        transfer_size=timings["transferSize"],
-        transfer_start_timings=["requestStart"],
-        tls_time=timings["secureConnectionStart"],
-        dns_time=timings["domainLookupEnd"],
-        connect_time=timings["connectEnd"],
-        total_time=timings["duration"],
-        effective_url=timings["name"],
-    )
+    def _browser_has_timing_capabilities(self):
+        return self._context.browser.capabilities["browserName"] != "chrome"
+
+    def _is_using_tls(self, name):
+        return name.startsWith("https")
+
+    def _get_tls_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["connectEnd"] - timings["secureConnectionStart"]
+        else:
+            logger.warning("Secure TLS connection not used, defaulting tls_time to 0")
+            return 0
+
+    def _get_tcp_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["secureConnectionStart"] - timings["connectStart"]
+        else:
+            return timings["connectEnd"] - timings["connectStart"]
+
+    def _send_page_load_metrics(self):
+        if self._browser_has_timing_capabilities():
+            timings, *_ = self._context.browser.execute_script(
+                "return performance.getEntriesByType('navigation')"
+            )
+            protocol = timings["nextHopProtocol"]
+            if protocol != "http/1.1":
+                logger.warning(
+                    f"Timings may be inaccurate as protocol is not http/1.1: {protocol}"
+                )
+            metrics = {
+                "tcp_time": self._get_tcp_timing(timings),
+                "tls_time": self._get_tls_time_if_applicable(timings),
+                "dns_lookup_time": timings["domainLookupEnd"]
+                - timings["domainLookupStart"],
+                "time_to_first_byte": timings["responseStart"] - timings["connectionEnd"],
+                "time_to_last_byte": timings["responseEnd"] - timings["connectionEnd"],
+                "page_weight": timings["transferSize"],
```

Comment:
We have some dom timings we can put in here as well.  I think some subset of:
- domInteractive - requestStart (total time taken to load the page from an app perspective, i.e. excluding TCP and TLS negotiation)
- domInteractive (total time to interactive)
- domInteractive - responseEnd (additional time taken after the network finishes to prep the page -- incentivizes e.g. async loading of scripts so you don't block until the whole page is loaded to start loading additional resources)
- domContentLoadedEventEnd - domContentLoadedEventStart (time spent in JS attached to the onload event.  It's nice that we have timing for this specifically, and I think a good chunk of JS stuff does tend to hang off this event.  However it's partially an architectural question: you could probably build an app that puts its processing elsewhere, to the usefulness will depend on the specifics of the design)
- There are also metrics for the unload event fired in the previous page (before navigating).  But that's even more architecture-dependent and less likely to be useful (and I don't think we need to bother with it, probably...)

Reply:
I've added these suggested metrics to the list we collect, however, the namings might need reviewing.